text/0000-classic-decorator.md

```diff
@@ -0,0 +1,196 @@
+- Start Date: 2019-03-14
+- Relevant Team(s): Ember.js, Ember Data, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/468
+- Tracking:
+
+# `@classic` Decorator
+
+## Summary
+
+Add a set of warnings for users who adopt native class syntax with
+`EmberObject` base classes, and a `@classic` decorator which can be used to
+disable these warnings on a per-class basis.
+
+## Motivation
+
+As we've added native class support to Ember and moved forward with Ember
+Octane, it's become increasingly clear that there are still a number of strange
+behaviors, edge-cases, and other issues with adopting native class syntax and
+extending from `EmberObject`. These issues include:
+
+- The timings of `init` and `constructor`, and how the two can interact through
+  a class hierachy in a way that causes confusing failures.
+- The differences between class fields and properties, and how class fields can
+  override properties even in a subclass.
+- The use of Mixins, which lack a native class alternative, and can force users
+  to interop and have even more failure cases.
+
+Despite these edge cases, the community has made progress using native class
+syntax, and it is absolutely necessary for some users (`ember-cli-typescript`
+users in particular). The ability to transition progressively to a new syntax is
+also very valuable, and allowing users to adopt native class syntax _without_
+completely rewriting their classes means that they can upgrade one step at a
+time.
+
+However, as we begin to move into a world where more and more classes _don't_
+extend from `EmberObject` (starting with `GlimmerComponent`), these edge cases
+will become more common and confusing. It won't be enough to rely on simple
+rules of thumb like "always use `init`" because in some cases, there won't _be_
+an `init` method. It'll also become a question of "which class am I extending?
+`GlimmerComponent`? `EmberComponent`?
+
+This RFC proposes adding a `@classic` decorator which can be used on classes
+that still use `EmberObject` APIs:
+
+```js
+@classic
+export default class ApplicationController extends Controller {
+  init() {
+    super.init(...arguments);
+    // setup the controller class
+  }
+}
+```
+
+This decorator will:
+
+1. Provide a visual hint to users that they class they are working on uses
+   `EmberObject` specific functionality, and to be aware that they are not
+   working with "just" native classes.
+2. Provide a hint to _linters_ so that they can use a different set of rules for
+   "classic" classes, such as always use `init`, etc. and native classes, such
+   as to never use `.extend()` or mixins.
+3. Disable a number of warnings that will alert users when they have encountered
+   an edge case, such as using `init` in a parent class and `constructor` in a
+   child class.
+
+This should allow users who wish to adopt native class syntax to do so
```

Comment:
should this also be a code mod? to add the @classic decorator to each EmberObject class?

Reply:
Absolutely! I think a codemod for this would be very helpful

Reply:
I'm not so sure. Adding @classic is fairly quick and it might behoove developers to evaluate whether each class needs it manually.